### ISSUE TITLE
Difficulty Reproducing Downstream Task Finetune Results with HF Trainer

### ISSUE THREAD

**Issue by yangzhao1230** [kuleshov-group/caduceus#37]
Due to my unfamiliarity with using PyTorch Lightning (PL), I attempted to use the Hugging Face (HF) Trainer for pretraining and fine-tuning downstream tasks.

1. I was able to largely reproduce the results from your paper on pretraining. For example, the following figure shows the training curve for reproducing caduceus-ph. (Although using the original learning rate of 8e-3 was more unstable, I opted to use a learning rate of 1e-4).
![image](https://github.com/kuleshov-group/caduceus/assets/52462696/00eab08c-d293-4635-bfe3-61d34dbf259b)

2. However, I encountered significant difficulty when trying to fine-tune downstream tasks using HF. Fine-tuning NT benchmarks with HF is theoretically straightforward, and although my actual implementation differs, you can refer to my self-contained Colab: [Colab Link](https://colab.research.google.com/drive/1lskCcitEZ4Js2eLnwb-CQkG53cgQk3VZ?usp=sharing). This Colab is mainly based on the fine-tuning tutorial provided by NT. When fine-tuning downstream tasks, the numerical results are highly unstable, often resulting in NaN losses, and I am completely unable to reproduce the results from the paper. The following figure shows my reproduction of the H4ac results. As you can see, with warmup_steps=500 and a learning rate of 1e-3, I can mostly reproduce the results from the paper (~0.6). 
![image](https://github.com/kuleshov-group/caduceus/assets/52462696/61a39cbd-8f99-46f2-8264-66f2fcc6b749)

However, in most cases, the results are very unstable. You can observe in the training loss below that the curves representing a loss of 0 indicate occurrences of NaN.
![image](https://github.com/kuleshov-group/caduceus/assets/52462696/27b1d812-adca-4514-b864-5f500880c78f)

3. I used `AutoModelForSequenceClassification` to load the initial weights you provided **without adding any conjoin-related parameters**. I am wondering if this is the main reason affecting performance, and **I am also curious about the impact of conjoin**. I did not seem to find related experimental explanations in the paper.

4. When running your PL-based code, I can mostly reproduce the results, but since the HF code would be much simpler, I would like to understand why my implementation fails to reproduce the results completely. I hope to get your help!

Thank you for your assistance!

**Comment by yangzhao1230** [comment]
In conclusion, I have two main questions:

1. Can the **HF Trainer** theoretically be used for pretraining and fine-tuning your model? Using the HF Trainer would **significantly simplify the code**.
2. Does directly loading the basic model (**without any conjoin-related parameters and reverse complement augmentations**) have a significant impact on downstream tasks?

**Comment by yangzhao1230** [comment]
We can observe a very strange phenomenon. The figure shows different learning rates: 1e-3, 5e-4, 1e-4, and 5e-5. Surprisingly, it is the intermediate value of 5e-5 that shows instability.
![image](https://github.com/kuleshov-group/caduceus/assets/52462696/90ce9ec9-e213-46ca-93da-d85a097bce02)

**Comment by yangzhao1230** [comment]
Sorry, I may need to confirm the issue of numerical instability again. I'm not sure if some random issues in the environment are causing occasional occurrences of `{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'epoch': 0.11}` situations, which do not seem to be related to the specific learning rate set. Therefore, I just need you to help me understand whether RC augmentation in downstream tasks is critical.

**Comment by yangzhao1230** [comment]
This is a simple experiment to illustrate this issue. Each time I initialize an AutoModelForSequenceClassification and a Linear layer, I notice that the values in the linear layers of AutoModelForSequenceClassification are very unusual, whereas the values in the Linear layer are normal.
![image](https://github.com/kuleshov-group/caduceus/assets/52462696/a9f46499-5914-4a3d-9272-4347e658b458)

**Comment by yangzhao1230** [comment]
Here I provide my experiment code. I tried changing the hardware device but still observed the same phenomenon.

```
import torch
import torch.nn as nn

from transformers import AutoModelForSequenceClassification

MAX_TRAILS = 10


for i in range(MAX_TRAILS):
    model = AutoModelForSequenceClassification.from_pretrained(
        "kuleshov-group/caduceus-ph_seqlen-1k_d_model-256_n_layer-4_lr-8e-3", trust_remote_code=True)
    
    # print model.score's max and min value 
    score_max_value = model.score.weight.max().item()
    score_min_value = model.score.weight.min().item()
    print(f"Trail {i}: Max value: {score_max_value}, Min value: {score_min_value}")

    # randomly initialize a linear with the same shape as model.score
    linear = nn.Linear(model.score.in_features, model.score.out_features, bias=False)
    # print linear's max and min value
    linear_max_value = linear.weight.max().item()
    linear_min_value = linear.weight.min().item()
    print(f"Trail {i}: Linear Max value: {linear_max_value}, Min value: {linear_min_value}")
```

**Comment by yair-schiff** [comment]
Perhaps there is an issue in initializing the linear layers of `CaduceusForSequenceClassification`. Did you try manually initializing these, or not using this class and simply adding `nn.Linear` layers as you have above?

**Comment by yangzhao1230** [comment]
Yes. I tried to re-initialize the weight and it worked.

**Comment by yangzhao1230** [comment]
I am still unsure about the exact cause of the bug. Given that the hyenadna codebase does not appear to have such issues, could you provide some suggestions on how to effectively identify and fix initialization bugs?

**Comment by yair-schiff** [comment]
I am also not sure why HF is not initializing the params of the linear prediction head in a standard way. In any case, I pushed a fix for this where I explicitly init these weights. See [here](https://github.com/kuleshov-group/caduceus/blob/f05031fb760afda342d95b926a943b37dcdde97e/caduceus/modeling_caduceus.py#L500-L505), for example. This should hopefully resolve the issue.

**Comment by ForestsKing** [comment]
Thank you for your work. It has successfully solved the problem that had been bothering me. Yesterday, after I downloaded the model weights from Hugging Face, I directly fine-tuned and tested them on downstream tasks. However, this morning, I noticed that the loss remained stuck at 2E+29, and the MCC was consistently zero. After integrating your corrected code, the loss seems to have returned to normal, and I’m really looking forward to seeing how Caduceus performs next.  

If time permits, I would also like to suggest synchronizing these code corrections to the Hugging Face repository. This way, fewer people will end up wasting time and resources like I did, and Caduceus can truly be used "out of the box" for everyone.
